server/api_archive.go

```diff
@@ -0,0 +1,126 @@
+package server
+
+// This is a low-level archiver which zips and sends GCA server persistent file
+// data using a POST command. This is intended to be used by an external service
+// to provide archival snapshots of raw data in the system. Only public data is
+// provided, with private keys zeroed before sending.
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/glowlabs-org/gca-backend/glow"
+)
+
+var (
+	ApiArchiveRateLimiter *glow.RateLimiter // Provide a rate limiter for this API.
+)
+
+func init() {
+	ApiArchiveRateLimiter = glow.NewRateLimiter(apiArchiveLimit, apiArchiveRate)
+}
+
+// ArchiveHandler provides the POST /archive api endpoint. It returns
+// uninterpreted file data intended to be used by an external service
+// to produce an archive.
+func (gcas *GCAServer) ArchiveHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		http.Error(w, "Only POST method is allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	if r.ContentLength != 0 {
+		http.Error(w, "Request should not have a body", http.StatusBadRequest)
+		return
+	}
+
+	if !ApiArchiveRateLimiter.Allow() {
+		http.Error(w, "Too many requests", http.StatusTooManyRequests)
+		return
+	}
+
+	buffer := new(bytes.Buffer)
+	zipw := zip.NewWriter(buffer)
+
+	files := []string{"server.keys", "gcaTempPubKey.dat", "gcaPubKey.dat", "equipment-authorizations.dat", "allDeviceStats.dat", "equipment-reports.dat"}
+
+	for _, name := range files {
+		err := gcas.addFile(name, zipw)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Error zipping file %v: %v", name, err), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	if err := zipw.Close(); err != nil {
+		if _, err := w.Write(buffer.Bytes()); err != nil {
+			http.Error(w, "Failed to close zip writer", http.StatusInternalServerError)
+			return
+		}
+	}
+
+	w.Header().Set("Content-Type", "application/zip")
+
+	if _, err := w.Write(buffer.Bytes()); err != nil {
+		http.Error(w, "Failed to write archive response", http.StatusInternalServerError)
+	}
+}
+
+func (gcas *GCAServer) addFile(name string, zipw *zip.Writer) error {
+	// To avoid races with any components writing to the file, lock up here.
+	gcas.mu.Lock()
+	defer gcas.mu.Unlock()
```

Comment:
You should never hold a mutex while doing network or disk I/O, this strategy doesn't work

Reply:
This implementation will cause the entire server to lock up every time someone calls the /archive endpoint, potentially for a minute or more

Reply:
Probably the best thing to do here is not lock the file. Then we'll have to go through the code and make sure all writes to the files that we have are complete. Meaning, any update to the file that needs to happen occurs in a single call to file.Write() or file.WriteAt(). I'm pretty sure the server is already designed this way but it wasn't intentional so we need to double check anyway.

We also need to extend the readme to explain that the server archives itself by grabbing whole files off of the disk rather than by putting together all of its state into a single archive file.

Reply:
I'm a bit wary because I know that using this as a backup strategy is going to put a lot of constraints on future development, but I actually think everything is scoped narrowly enough that it won't matter much

Reply:
I will have a look at the places where we write files. We could also choose to use https://linux.die.net/man/2/flock shared and exclusive file locking for this purpose.

Reply:
It makes sense that in an append only system a file lock is overkill.

Reply:
I confirmed the following single writes, and made this ordering

loadGCAServerKeys server.keys Write
saveGCAKey gcaPubKey.dat WriteFile
saveAllDeviceStats allDeviceStats.dat Write
saveEquipment equipment-authorizations.dat Write
saveEquipmentReport equipment-reports.dat Write